src/compiler/checker.ts

```diff
@@ -6645,6 +6642,41 @@ module ts {
                     callIsIncomplete = !!templateLiteral.isUnterminated;
                 }
             }
+            else if (node.kind === SyntaxKind.Decorator) {
+                let decorator = <Decorator>node;
+                switch (decorator.parent.kind) {
+                    case SyntaxKind.ClassDeclaration:
+                    case SyntaxKind.ClassExpression:
+                        // A class decorator will have one argument (see `ClassDecorator` in core.d.ts)
+                        adjustedArgCount = 1;
+                        typeArguments = undefined;
+                        break;
+
+                    case SyntaxKind.PropertyDeclaration:
+                        // A property declaration decorator will have two arguments (see 
+                        // `PropertyDecorator` in core.d.ts)
+                        adjustedArgCount = 2;
+                        typeArguments = undefined;
+                        break;
+
+                    case SyntaxKind.MethodDeclaration:
+                    case SyntaxKind.GetAccessor:
+                    case SyntaxKind.SetAccessor:
+                        // A method or accessor declaration decorator will have two or three arguments (see 
+                        // `MethodDecorator` in core.d.ts)
+                        adjustedArgCount = signature.parameters.length >= 3 ? 3 : 2;

```

Comment:
Intuitively it seems like it would work to just make `descriptor` optional in core.d.ts. But you actually want the dual of an optional parameter, something more like an "optional argument". This is a weird case because usually the number of arguments is determined by the call, and optional parameters are used to expand/contract the signature as necessary. Here it's the reverse, the signature is determined by what the user wrote, and we have this "optional argument" that expands and contracts the call to match the arity of the signature. Do you think you could write a quick note to that effect as a comment right here?